src/blockstore/insert_shred.zig

```diff
@@ -0,0 +1,1890 @@
+            if (try self.db.get(schema.erasure_meta, erasure_set)) |meta_| {
+                erasure_meta_entry.value_ptr.* = .{ .clean = meta_ };
+            } else {
+                std.debug.assert(erasure_metas.remove(erasure_set));
+            }
+        }
+
+        return newly_completed_data_sets;
+    }
+
+    /// agave: get_index_meta_entry
+    fn getIndexMetaEntry(
+        self: *Self,
+        allocator: std.mem.Allocator,
+        slot: Slot,
+        working_set: *AutoHashMap(Slot, IndexMetaWorkingSetEntry),
+        index_meta_time_us: *u64,
+    ) !*IndexMetaWorkingSetEntry {
+        // TODO: redundant get or put pattern
+        var timer = try Timer.start();
+        const entry = try working_set.getOrPut(slot);
+        if (!entry.found_existing) {
+            if (try self.db.get(schema.index, slot)) |item| {
+                entry.value_ptr.* = .{ .index = item };
+            } else {
+                entry.value_ptr.* = IndexMetaWorkingSetEntry.init(allocator, slot);
+            }
+        }
+        index_meta_time_us.* += timer.read().asMicros();
+        return entry.value_ptr;
+    }
+
+    /// agave: get_slot_meta_entry
+    fn getSlotMetaEntry(
+        self: *Self,
+        working_set: *AutoHashMap(u64, SlotMetaWorkingSetEntry),
+        slot: Slot,
+        parent_slot: Slot,
+    ) !*SlotMetaWorkingSetEntry {
+        // TODO: redundant get or put pattern
+        const entry = try working_set.getOrPut(slot);
+        if (!entry.found_existing) {
+            const sm = try self.allocator.create(SlotMeta);
+            if (try self.db.get(schema.slot_meta, slot)) |item| {
+                sm.* = item;
+                entry.value_ptr.* = .{
+                    .new_slot_meta = sm,
+                    .old_slot_meta = try item.clone(self.allocator),
+                };
+            } else {
+                sm.* = SlotMeta.init(self.allocator, slot, parent_slot);
+                entry.value_ptr.* = .{
+                    .new_slot_meta = sm,
+                };
+            }
+        }
+        return entry.value_ptr;
+    }
+
+    /// agave: insert_coding_shred
+    fn insertCodingShred(
+        index_meta: *meta.Index,
+        shred: CodingShred,
+        write_batch: *WriteBatch,
+    ) !void {
+        const slot = shred.fields.common.slot;
+        const shred_index: u64 = @intCast(shred.fields.common.index);
+
+        assertOk(shred.sanitize());
+
+        try write_batch.put(schema.code_shred, .{ slot, shred_index }, shred.fields.payload);
+        try index_meta.code.put(slot);
+    }
+
+    /// Check if the shred already exists in blockstore
+    /// agave: is_data_shred_present
+    fn isDataShredPresent(
+        shred: DataShred,
+        slot_meta: *SlotMeta,
+        data_index: *meta.ShredIndex,
+    ) bool {
+        const shred_index: u64 = @intCast(shred.fields.common.index);
+        return shred_index < slot_meta.consumed or data_index.contains(shred_index);
```

Comment:
> reading more into this - would n_consecutive_shreds_recieved be a more accurate name than consumed?

I agree the name could be improved. Another key aspect to consider is that they need to be consecutive shreds that were received *starting at zero*.

55e66ef04e23e63a5fc4fcc4fd3f5bcd8f66455a

> also since this is the only field we use in slot_meta, should we just pass in the u64 directly?

That's definitely an option. It's nice how it minimizes the dependencies of this function. But I think there are also advantages to having SlotMeta as the parameter here.

What's actually the point of this function? You could just completely inline and remove this function, and it wouldn't be a big deal. My though process is that it has the responsibility of figuring out where to look in the metadata to figure out if the shred is present, and it abstracts that from the calling context. So it seems reasonable for it to work with the metadata types. 

Another way to look at this is that SlotMeta is a strong type that ensures the correct data is getting into the function. Taking a u64 here means it's easier to make a mistake and pass in the wrong data. Then you need to make sure that two contexts (this and the calling scope) both need to understand which integer is needed here.

I don't have a strong feeling about this function, so let me know if you specifically want it a certain way.

> but also - would there ever be the case that one of the or are true but not the other - seems like they both state the same thing (feels almost redundant)

Let's say you've received shreds 0,1,2,7,8,9. Shred 0,1,2 will be true for both checks but 7,8,9 will only be true for the second check because 3-6 are missing. The idea here is that it starts with the less expensive check (less than), and if that doesn't give a definitive answer, then it falls back to the more expensive but definitive check (hashmap lookup).